src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,274 @@
+package duke.parser;
+
+import java.time.DateTimeException;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+
+import duke.DukeException;
+import duke.command.AddCommand;
+import duke.command.ByeCommand;
+import duke.command.Command;
+import duke.command.DateCommand;
+import duke.command.DeleteCommand;
+import duke.command.FindCommand;
+import duke.command.ListCommand;
+import duke.command.MarkCommand;
+import duke.command.UnmarkCommand;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDo;
+
+/**
+ * Provides static methods to parse user queries and data file.
+ */
+public class Parser {
+    private static String emptyDeadlineDate = "Please specify deadline date using /by\n"
+        + "e.g. deadline report /by 2023-12-31";
+    private static String emptyEventDate = "Please specify event start and end dates using /from and /to\n"
+        + "e.g. event holiday /from 2023-06-01 /to 2023-06-30";
+    private static String invalidDate = "Please provide date with the following format: YYYY-MM-DD";
+    private static String invalidEndDate = "Your end date is before start date";
+
+    /**
+     * Converts a String date from YYYY-MM-DD to d MMM yyyy format.
+     * @param dateStr String date of YYYY-MM-DD format
+     * @return String date of d MMM yyyy format
+     * @throws DukeException if invalid date
+     */
+    public static String convertToDmy(String dateStr) throws DukeException {
+        try {
+            LocalDate date = LocalDate.parse(dateStr);
+            return date.format(DateTimeFormatter.ofPattern("d MMM yyyy"));
+        } catch (DateTimeException e) {
+            throw new DukeException(invalidDate);
+        }
+    }
+
+    /**
+     * Converts any string separated by white spaces to List of String.
+     * @param s String separated by white spaces
+     * @return List of String
+     */
+    private static List<String> convertToList(String s) {
+        String[] queries = s.trim().split("\\s+");
+        List<String> queryList = Arrays.asList(queries);
+        return queryList;
+    }
+
+    /**
+     * Parses each line of String in data file into Task.
+     * @param s Line of String in data file
+     * @return Task
+     * @throws DukeException if String cannot be parsed due to invalid
+     *     format or invalid date
+     */
+    public static Task parseFile(String s) throws DukeException {
+        String[] q = s.trim().split(",>");
+        try {
+            switch (q[0]) {
+            case "deadline":
+                return new Deadline(q[1], q[2].equals("X"), LocalDate.parse(q[3]));
+            case "event":
+                return new Event(q[1], q[2].equals("X"), LocalDate.parse(q[3]), LocalDate.parse(q[4]));
+            case "todo":
+                return new ToDo(q[1], q[2].equals("X"));
+            default:
+                throw new DukeException("Error parsing file data");
```

Comment:
Perhaps you could create more exceptions that inherit from `DukeException` such as `FileParseException`!